src/node/config.ts

```diff
@@ -627,106 +624,75 @@ async function loadConfigFromBundledFile(
   return config
 }
 
-function resolvePlugin(config: UserConfig, plugin: Plugin): UserConfig {
-  return {
-    ...config,
-    ...plugin,
-    alias: {
-      ...plugin.alias,
-      ...config.alias
-    },
-    define: {
-      ...plugin.define,
-      ...config.define
-    },
-    transforms: [...(config.transforms || []), ...(plugin.transforms || [])],
-    indexHtmlTransforms: [
-      ...(config.indexHtmlTransforms || []),
-      ...(plugin.indexHtmlTransforms || [])
-    ],
-    resolvers: [...(config.resolvers || []), ...(plugin.resolvers || [])],
-    configureServer: ([] as ServerPlugin[]).concat(
-      config.configureServer || [],
-      plugin.configureServer || []
-    ),
-    configureBuild: ([] as BuildPlugin[]).concat(
-      config.configureBuild || [],
-      plugin.configureBuild || []
-    ),
-    vueCompilerOptions: {
-      ...config.vueCompilerOptions,
-      ...plugin.vueCompilerOptions
-    },
-    vueTransformAssetUrls: mergeAssetUrlOptions(
-      config.vueTransformAssetUrls,
-      plugin.vueTransformAssetUrls
-    ),
-    vueTemplatePreprocessOptions: {
-      ...config.vueTemplatePreprocessOptions,
-      ...plugin.vueTemplatePreprocessOptions
-    },
-    vueCustomBlockTransforms: {
-      ...config.vueCustomBlockTransforms,
-      ...plugin.vueCustomBlockTransforms
-    },
-    rollupInputOptions: mergeObjectOptions(
-      config.rollupInputOptions,
-      plugin.rollupInputOptions
-    ),
-    rollupOutputOptions: mergeObjectOptions(
-      config.rollupOutputOptions,
-      plugin.rollupOutputOptions
-    ),
-    enableRollupPluginVue:
-      config.enableRollupPluginVue || plugin.enableRollupPluginVue
+function mergePlugin(config: ResolvedConfig, plugin: Plugin) {
+  if (typeof plugin === 'function') {
+    plugin = plugin(config)
```

Comment:
I removed `changeConfig` in favor of plugin functions that receive a stable reference to the `config` object.